### PR TITLE
fix: Match promisified fs.read call return with original implementation

### LINKED ISSUE #2127: [Bug] Promisified fs.read calls return a different result when run with Yarn PnP than through Node
- [x] I'd be willing to implement a fix

**Describe the bug**

Promisified `fs.read` returns the number of bytes read instead of an object containing the `buffer` and the `bytesRead`.

**To Reproduce**

https://github.com/Mike-Dax/promisified-fs-read-bug

```
// test.js
const fs = require("fs");
const { promisify } = require("util");
const fsReadAsync = promisify(fs.read);

const path = require("path");

async function main() {
  // Setup the file
  const p = path.join(__dirname, "./test.txt");
  fs.writeFileSync(p, "hey");

  // Read the file
  const fd = fs.openSync(p);
  const buffer = Buffer.alloc(16);
  const ret = await fsReadAsync(fd, buffer, 0, 4, null);

  console.log(ret);

  console.log(buffer.toString());
}

main();
```

The results are:

`yarn node test.js`

```
3
hey
```

`node test.js`

```
{
  bytesRead: 3,
  buffer: <Buffer 68 65 79 00 00 00 00 00 00 00 00 00 00 00 00 00>
}
hey
```

**Additional context**

This seems to be the root cause of https://github.com/serialport/node-serialport/issues/2158

### PR DESCRIPTION
**Pull request by Mike-Dax**
**What's the problem this PR addresses?**

Fixes #2127
Fixes https://github.com/serialport/node-serialport/issues/2158

...

**How did you fix it?**

Node stores whether or not the promisified version should have a different return shape in a symbol (`internalUtil.customPromisifyArgs`) that's private, we can't access that.

https://github.com/nodejs/node/blob/dc79f3f37caf6f25b8efee4623bec31e2c20f595/lib/fs.js#L559-L560

However the util.promisify function caches implementations in a cache key that's a global symbol (available at `util.promisify.custom`). We can pre-populate that cache with a version of the function that works.

https://github.com/nodejs/node/blob/ba684805b6c0eded76e5cd89ee00328ac7a59365/lib/internal/util.js#L274,L281

This PR implements a pre-cached function that returns an object of the same shape as `util.promisify(fs.read)` would on a standard node system
...

**Checklist**

- [x] I have read the [Contributing Guide](https://yarnpkg.com/advanced/contributing).

- [x] I have set the packages that need to be released for my changes to be effective.

- [x] I will check that all automated PR checks pass before the PR gets reviewed.

### CODE REVIEW

**Review by merceyz — CHANGES_REQUESTED** [review]
Could you add a test to make sure we don't regress on this in the future?

**Comment by arcanis** [comment]
For that, a unit test in `patchedFs.test.ts` will do the trick

File: packages/yarnpkg-fslib/sources/index.ts [hunk 1]
```diff
@@ -284,6 +285,19 @@ export function patchFs(patchedFs: typeof fs, fakeFs: FakeFS<NativePath>): void
...
   }
+
+  /** util.promisify implementations */
+  {
+    // Override the promisified version of `fs.read` to return an object as per
+    // https://github.com/nodejs/node/blob/dc79f3f37caf6f25b8efee4623bec31e2c20f595/lib/fs.js#L559-L560
+    // and
+    // https://github.com/nodejs/node/blob/ba684805b6c0eded76e5cd89ee00328ac7a59365/lib/internal/util.js#L293
+    // @ts-expect-error
+    patchedFs.read[promisify.custom] = async (p: number, buffer: Buffer, ...args: Array<any>) => {
```
Comment (arcanis) [inline packages/yarnpkg-fslib/sources/index.ts:296]: Shouldn't we do the same for `write` / `readv` / `writev`?

Comment (Mike-Dax) [reply packages/yarnpkg-fslib/sources/index.ts:296]: Those other three aren't patched, so this problem doesn't occur.

If we wanted to patch those methods, then yeah we should do the promisify patching.

**Comment by Mike-Dax** [comment]
I've added a test to validate the returned object shape. I tried to write a test that ran the unpatched `fs.read` then patched `fs`, and compared the two, but I couldn't seem to get access to the unpatched `fs.read`.

**Pull request merged**